src/command.ml

```diff
@@ -148,15 +148,42 @@ let string_of_command_spec_with_calls call_with_tags call_with_target resolve_vi
       | P p -> put_space (); put_filename p
       | Px u -> put_space (); put_filename u; call_with_target u
       | V v -> if resolve_virtuals then do_spec (virtual_solver v)
-               else (put_space (); Printf.bprintf b "<virtual %s>" (Shell.quote_filename_if_needed v))
-      | S l -> List.iter do_spec l
+              else (put_space (); Printf.bprintf b "<virtual %s>" (encode v))
+      | S l -> do_specs l
       | T tags -> call_with_tags tags; do_spec (!tag_handler tags)
-      | Quote s -> put_space (); put_filename (aux s)
+      | Quote s -> put_space (); put_filename (aux encode s)
+    and do_specs = function
+      | [] -> ()
+      | (A "-pp" as pp) :: Quote q :: rest ->
```

Comment:
It feels very fishy to me to special-case `-pp` for special quoting rule. Why not apply this logic to all `Quote` arguments?

Reply:
Should we change the definition of Quote ?
It's currently 
>     | Quote of spec  (** A string that should be quoted like a
>                           filename but isn't really one. *)

Reply:
My understanding is that `Quote` is in fact used for arguments that are themselves commands or prefixes of command, meant to be passed to the shell at some point. We could indeed change the (fairly unclear) documentation to reflect this.